packages/json-api/src/Resource.ts

```diff
@@ -0,0 +1,265 @@
+};
+
+type ResourceAttributeHelpers = Array<[string, ResourceAttributeMapper]>;
+
+const defaultIdMapper: ResourceIdMapper = {
+  attr: "id",
+  format(value) {
+    if (value === null) {
+      return null;
+    }
+    if (value === undefined) {
+      return undefined;
+    }
+    return String(value);
+  },
+  parse(resourceObj) {
+    return resourceObj.id != null ? String(resourceObj.id) : null;
+  }
+};
+
+const defaultAttributeMapper: ResourceAttributeMapper = {
+  getter: (data: JSONObject, key: string) => data[key] as JSONObject,
+  formatter: (value: JSONValue) => value
+};
+
+type ResourceRelationshipMapper = (
+  data: JSONObject,
+  key: string
+) => JSONAPIResourceRelationship;
+
+type ResourceRelationshipHelpers = Array<[string, ResourceRelationshipMapper]>;
+
+export default class ResourceImpl implements Resource {
+  private registry: Registry;
+
+  private type: string;
+
+  private idSpec: ResourceIdMapper;
+
+  private attributes: ResourceAttributeHelpers;
+
+  private relationships: ResourceRelationshipHelpers | null;
+
+  constructor(registry: Registry, type: string, spec: ResourceSpec) {
+    this.registry = registry;
+    this.type = type;
+    this.idSpec = { ...defaultIdMapper, ...spec.id };
+    this.attributes = this.normalizeAttributes(spec.attributes);
+    this.relationships = this.normalizeRelationships(
+      registry,
+      spec.relationships
+    );
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  private normalizeAttributes(
+    specs?: ResourceAttributesSpec
+  ): ResourceAttributeHelpers {
+    if (Array.isArray(specs)) {
+      return specs.map(spec => {
+        if (typeof spec === "string") {
+          return [spec, defaultAttributeMapper];
+        }
+        return spec;
+      });
+    }
+    if (specs != null) {
+      return Object.entries(specs).map(([key, value]) => [
+        key,
+        { ...defaultAttributeMapper, ...value }
+      ]);
+    }
+    return [];
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  private normalizeRelationships(
+    registry: Registry,
+    relationshipSpecs?: ResourceRelationshipsSpec
+  ): ResourceRelationshipHelpers | null {
+    if (relationshipSpecs != null) {
+      return Object.entries(relationshipSpecs).map(([key, value]) => {
+        // eslint-disable-next-line no-shadow
+        const getter = (data: JSONObject, key: string) => {
+          let relatedResourceClass: Resource | null;
+          if (typeof value === "string") {
+            relatedResourceClass = registry.find(value);
+          } else if (value instanceof ResourceImpl) {
+            relatedResourceClass = value;
+          } else {
+            relatedResourceClass = registry.find(
+              (value as { type: string }).type
+            );
+          }
+
+          if (relatedResourceClass == null) {
+            throw new Error(`Cannot find ${key} resource`);
+          }
+
+          // eslint-disable-next-line no-underscore-dangle
```

Comment:
So many `eslint-disable-next-line` is disturbing

Reply:
I got rid of almost all of them. This one is actually needed, because `_embed` field is a part of library API.